### PR TITLE
[posix] add Backbone Router multicast routing for non-Linux platforms

### PR DESCRIPTION
**Pull request by clara-mostly-human**
Follow-up to openthread/ot-br-posix#3590, where @jwhui asked for this to live here rather than in `ot-br-posix`.

`ot::Posix::MulticastRoutingManager` relies on the kernel's IPv6 multicast routing (MRT6) and `#error`s everywhere but Linux, so a posix Backbone Router on macOS or a BSD has no multicast forwarding. This PR adds an implementation for those platforms that forwards in userspace, behind the same public interface, so `InfraNetif` drives it exactly as it does today.

### How it works

- A BPF tap on the Thread interface delivers what the Thread stack hands to the host (received packets only); a second tap on the infrastructure interface delivers that link's multicast. A filter program on each keeps everything except multicast of admin-local scope or larger in the kernel.
- **Thread → backbone**: every packet that passes the policy (routable source, hop limit above 1) goes out in an Ethernet frame with the hop limit decremented — what the outbound forwarding cache entries do with MRT6.
- **Backbone → Thread**: only for groups in the Multicast Listener table (the existing `HasMulticastListener()`) — what the inbound forwarding cache entries do with MRT6. The packet is written to the tunnel interface.
- **Storm safety**: a packet is forwarded once. A duplicate cache covers both directions, which also stops the forwarder from picking up its own emissions on the backbone tap. Per-group and overall token buckets apply in each direction, tighter into the mesh, where every packet is flooded by MPL. All tables are fixed-size.

### Layout

As discussed on the issue, the non-Linux implementation is a separate source file and the private section of the header is split by platform:

- `multicast_routing.cpp` keeps what both implementations share (`SetUp`/`TearDown`, enabling on Primary with retries, the MLD join on the infrastructure interface for registered groups, `HasMulticastListener()`, `HandleStateChange()`). The MRT6 code moves under `#ifdef __linux__` **unchanged** — the diff of that file is moves plus guards, and the constructor coming over from the header.
- `multicast_routing_bpf.cpp` — the non-Linux data plane.
- `bpf_tap.*` — a small wrapper around a BPF device (attach, filter, read, write).
- `multicast_forwarding.*` — the per-packet checks: policy, duplicate cache, rate limiter. No platform dependencies.

On non-Linux platforms `mMulticastRouterSock` is a plain UDP socket that only carries the group memberships on the infrastructure interface; joining there is also what makes the interface deliver the group to the tap.

### Testing

macOS 26 (arm64), `ot-daemon` as Primary BBR on a simulated RCP, an isolated bridge as the backbone, a simulated FTD on the mesh registered for `ff05::abcd`:

| | |
|---|---|
| mesh → backbone | 10 groupcasts → 10 on the backbone, 10 distinct, hop limit 64 → 63; realm-local (`ff03::`) groupcasts: 0 on the backbone |
| backbone → mesh | 10 groupcasts → 10 at the mesh listener, 10 distinct; a group nobody registered for: 0 |
| flood | 200 back-to-back from the backbone → 23 into the mesh, 177 rate limited; BR stays leader, node stays attached |
| no loop | the 15 packets forwarded to the backbone are seen again on the backbone tap and all 15 are dropped as duplicates |
| BBR gating | `bbr disable` → 0 forwarded; `bbr enable` → forwarding resumes |

Counters at the end of the run:

```
Thread -> Backbone: received 15, forwarded 15, rejected 0, no listener 0, duplicates 0, rate limited 0, errors 0
Backbone -> Thread: received 228, forwarded 33, rejected 0, no listener 3, duplicates 15, rate limited 177, errors 0
```

The same data plane (as the `ot-br-posix` version in openthread/ot-br-posix#3593) has been forwarding on my production border router, a Mac mini with an EFR32MG24 RCP, since 12 September.

The per-packet checks also pass a standalone test (policy verdicts, duplicate window and eviction order, per-group and overall limits, a saturated group not draining the shared budget). I did not find an existing home for posix platform unit tests; happy to add it as a gtest if you can point me at where you would like it.

Linux is unaffected by construction. On my fork this branch passes the POSIX workflow (Linux and macOS legs), Border Router, OTBR DinD, Unit, Nexus and the `pretty` check; the compiler matrix of the Build workflow was still running when I opened this.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/posix/platform/bpf_tap.hpp [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+ *   This file includes definitions for a packet tap on a network interface
+ *   through the BSD Packet Filter.
+ */
+
+#ifndef OT_POSIX_PLATFORM_BPF_TAP_HPP_
+#define OT_POSIX_PLATFORM_BPF_TAP_HPP_
+
+#include "openthread-posix-config.h"
+
+#ifndef __linux__
```
Comment (jwhui) [inline src/posix/platform/bpf_tap.hpp]: Guard with `#if OPENTHREAD_POSIX_CONFIG_BACKBONE_ROUTER_MULTICAST_ROUTING_ENABLE && !defined(__linux__)` (matching `multicast_routing_bpf.cpp`) so `BpfTap` isn't compiled when BBR multicast routing is disabled.

File: src/posix/platform/bpf_tap.cpp [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+    VerifyOrExit(!IsOpen(), error = OT_ERROR_ALREADY);
+
+    // The BPF device does not clone on every platform: take the first free unit.
+    for (uint16_t unit = 0; unit < kMaxUnits && mFd < 0; unit++)
+    {
+        char path[sizeof("/dev/bpf65535")];
+
+        snprintf(path, sizeof(path), "/dev/bpf%u", unit);
+        mFd = open(path, O_RDWR | O_CLOEXEC);
```
Comment (jwhui) [inline src/posix/platform/bpf_tap.cpp]: 1. Pass `O_NONBLOCK` to `open()`. Without it, `mFd` is blocking (`errno == EAGAIN` in `Read()` is unreachable) and a spurious `select()` wakeup will block the mainloop.
2. Consider trying `/dev/bpf `first before looping through `/dev/bpf0` .. `/dev/bpf255`.

File: src/posix/platform/bpf_tap.cpp [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+    strncpy(ifr.ifr_name, aIfName, sizeof(ifr.ifr_name) - 1);
+    VerifyOrExit(ioctl(mFd, BIOCSETIF, &ifr) == 0, error = OT_ERROR_FAILED);
+
+    // Deliver frames as they arrive rather than when the buffer fills.
+    value = 1;
+    VerifyOrExit(ioctl(mFd, BIOCIMMEDIATE, &value) == 0, error = OT_ERROR_FAILED);
+
+    // Frames written to the tap carry their own link-layer header.
+    value = 1;
+    VerifyOrExit(ioctl(mFd, BIOCSHDRCMPLT, &value) == 0, error = OT_ERROR_FAILED);
```
Comment (jwhui) [inline src/posix/platform/bpf_tap.cpp]: Setting `BIOCSHDRCMPLT = 1` on `DLT_NULL` breaks `BpfTap::Write()` on FreeBSD `tun`: `bpfwrite()` overwrites `dst.sa_family = AF_UNSPEC` with `pseudo_AF_HDRCMPLT`, causing `tunoutput()` to fail with `EAFNOSUPPORT`. Move `BIOCSHDRCMPLT` after `BIOCGDLT` and only enable it when `mDataLinkType == DLT_EN10MB`.

File: src/posix/platform/multicast_forwarding.hpp [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+ *   forwarder applies to every packet: forwarding policy, duplicate
+ *   suppression and rate limiting.
+ */
+
+#ifndef OT_POSIX_PLATFORM_MULTICAST_FORWARDING_HPP_
+#define OT_POSIX_PLATFORM_MULTICAST_FORWARDING_HPP_
+
+#include "openthread-posix-config.h"
+
+#ifndef __linux__
```
Comment (jwhui) [inline src/posix/platform/multicast_forwarding.hpp]: Remove `#ifndef __linux__` since this module has no platform dependencies. That allows running the unit tests on Linux CI as well (you can add your unit test to `tests/gtest/CMakeLists.txt` under `ot-ftd-gtest` or `tests/unit/CMakeLists.txt` via `ot_unit_test`).

File: src/posix/platform/multicast_forwarding.cpp [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "posix/platform/multicast_forwarding.hpp"
+
+#ifndef __linux__
```
Comment (jwhui) [inline src/posix/platform/multicast_forwarding.cpp]: Remove `#ifndef __linux__` since this module has no platform dependencies. That allows running the unit tests on Linux CI as well (you can add your unit test to `tests/gtest/CMakeLists.txt` under `ot-ftd-gtest` or `tests/unit/CMakeLists.txt` via `ot_unit_test`).

File: src/posix/platform/multicast_forwarding.cpp [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+}
+
+bool TokenBucket::Take(uint64_t aNow)
+{
+    bool     taken   = false;
+    uint64_t full    = mBurst * kScale;
+    uint64_t elapsed = (aNow > mLastRefill) ? (aNow - mLastRefill) : 0;
+
+    // One microsecond earns `mRatePerSecond` millionths of a token.
+    if (elapsed >= full / (mRatePerSecond ? mRatePerSecond : 1))
```
Comment (jwhui) [inline src/posix/platform/multicast_forwarding.cpp]: Guard the refill logic with `if (mRatePerSecond > 0)`. Currently, if `mRatePerSecond == 0`, `mTokens` still refills to `full` once `elapsed >= full`.

File: src/posix/platform/bpf_tap.cpp [hunk 7]
```diff
@@ -0,0 +1,211 @@
...
+    }
+
+    while (offset + sizeof(struct bpf_hdr) <= static_cast<size_t>(count))
+    {
+        struct bpf_hdr header;
+        size_t         frameOffset;
+
+        memcpy(&header, mBuffer + offset, sizeof(header));
+        frameOffset = offset + header.bh_hdrlen;
+        VerifyOrExit(frameOffset + header.bh_caplen <= static_cast<size_t>(count), error = OT_ERROR_PARSE);
```
Comment (jwhui) [inline src/posix/platform/bpf_tap.cpp]: Validate `header.bh_hdrlen >= sizeof(header)` and check `header.bh_caplen == header.bh_datalen && header.bh_caplen <= UINT16_MAX` before casting to `uint16_t` (avoiding 32-bit overflow in `frameOffset + header.bh_caplen` and dropping truncated captures).

File: src/posix/platform/multicast_forwarding.cpp [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+
+Verdict Check(const uint8_t *aPacket, uint16_t aLength)
+{
+    Verdict      verdict = kForward;
+    Ip6::Address destination;
+    Ip6::Address source;
+    uint16_t     payloadLength;
+
+    VerifyOrExit(aLength >= kIp6HeaderSize && (aPacket[0] >> 4) == 6, verdict = kNotIp6);
+    payloadLength = static_cast<uint16_t>((aPacket[kPayloadLengthOffset] << 8) | aPacket[kPayloadLengthOffset + 1]);
```
Comment (jwhui) [inline src/posix/platform/multicast_forwarding.cpp:56]: 1. **Ethernet padding breaks `DedupCache`**: `Check()` allows `kIp6HeaderSize + payloadLength <= aLength`. For small packets (`payloadLength < 6`, IPv6 length 40..45 bytes), Ethernet pads frames to 60 bytes (`aLength = 46`). A packet forwarded from Thread (`aLength = 40..45`) and read back on `mBackboneTap` (`aLength = 46`) will be fingerprinted over 46 bytes, missing `DedupCache` and looping back into Thread with trailing padding. Have `Check()` return/output `kIp6HeaderSize + payloadLength` and use that length for `Fingerprint()` and `Write()`.
2. Also reject loopback sources (`!source.IsLoopback()`).

File: src/posix/platform/multicast_routing_bpf.cpp [hunk 9]
```diff
@@ -0,0 +1,464 @@
...
+    , mState(kStateDisabled)
+    , mRetryIntervalMs(kMinRetryIntervalMs)
+    , mNextRetryTime(0)
+{
+    memset(mBackboneMac, 0, sizeof(mBackboneMac));
+    memset(&mThreadToBackbone, 0, sizeof(mThreadToBackbone));
+    memset(&mBackboneToThread, 0, sizeof(mBackboneToThread));
+}
+
+void MulticastRoutingManager::Add(const Ip6::Address &aAddress)
```
Comment (jwhui) [inline src/posix/platform/multicast_routing_bpf.cpp]: `Add()` and `Remove()` duplicate `multicast_routing.cpp` (minus `LogResult` and MFC calls). Consider keeping `Add()`/`Remove()` in `multicast_routing.cpp` with `#ifdef __linux__` around the MFC calls.

File: src/posix/platform/multicast_routing_bpf.cpp [hunk 10]
```diff
@@ -0,0 +1,464 @@
...
+    {
+        Mainloop::SetTimeoutIfEarlier((mNextRetryTime > now) ? (mNextRetryTime - now) : 0, aContext);
+        ExitNow();
+    }
+
+    VerifyOrExit(IsEnabled());
+
+    Mainloop::AddToReadFdSet(mThreadTap.GetFd(), aContext);
+    Mainloop::AddToReadFdSet(mBackboneTap.GetFd(), aContext);
+    Mainloop::SetTimeoutIfEarlier((mNextReportTime > now) ? (mNextReportTime - now) : 0, aContext);
```
Comment (jwhui) [inline src/posix/platform/multicast_routing_bpf.cpp]: Only call `Mainloop::SetTimeoutIfEarlier` for `mNextReportTime` when `mCountersChanged` is `true` so an idle forwarder doesn't wake the mainloop every 5 minutes.

File: src/posix/platform/multicast_routing_bpf.cpp [hunk 11]
```diff
@@ -0,0 +1,464 @@
...
+    {
+        error = mThreadTap.Read(HandleThreadFrame, this);
+    }
+
+    if (error == OT_ERROR_NONE && Mainloop::IsFdReadable(mBackboneTap.GetFd(), aContext))
+    {
+        error = mBackboneTap.Read(HandleBackboneFrame, this);
+    }
+
+    if (error != OT_ERROR_NONE)
```
Comment (jwhui) [inline src/posix/platform/multicast_routing_bpf.cpp]: Only tear down and retry on `error == OT_ERROR_FAILED`. On `OT_ERROR_PARSE`, `errno` is not set (`strerror(errno)` is stale) and tearing down the taps is unnecessary.

File: src/posix/platform/multicast_routing_bpf.cpp [hunk 12]
```diff
@@ -0,0 +1,464 @@
...
+    uint64_t     now = otPlatTimeGet();
+    Verdict      verdict;
+    Ip6::Address group;
+    uint8_t      frame[kMaxFrameSize];
+    uint16_t     frameLength;
+
+    mThreadToBackbone.mReceived++;
+    mCountersChanged = true;
+
+    verdict = Check(aPacket, aLength);
```
Comment (jwhui) [inline src/posix/platform/multicast_routing_bpf.cpp]: 1. **Missing Mesh-Local source check**: Matching Linux (`multicast_routing.cpp` line 395), reject packets where `GetSource(aPacket).GetPrefix() == AsCoreType(otThreadGetMeshLocalPrefix(gInstance))`. If the BBR host itself subscribes to a group, `Ip6::PassToHost` passes mesh-local-sourced packets to `sTunFd`, which `mThreadTap` would otherwise forward onto the backbone.
2. Check frame length bounds *before* `mDedupCache.Check()` and `mThreadToBackboneLimiter.Allow()` so rejected oversized packets don't consume tokens or dedup entries.

File: src/posix/platform/multicast_routing_bpf.cpp [hunk 13]
```diff
@@ -0,0 +1,464 @@
...
+    uint64_t     now = otPlatTimeGet();
+    Verdict      verdict;
+    Ip6::Address group;
+    uint8_t      frame[kMaxFrameSize];
+    uint32_t     family = AF_INET6;
+
+    mBackboneToThread.mReceived++;
+    mCountersChanged = true;
+
+    verdict = Check(aPacket, aLength);
```
Comment (jwhui) [inline src/posix/platform/multicast_routing_bpf.cpp]: 1. **Thread MTU / `sTunFd` truncation**: `sizeof(frame)` allows up to 1514-byte packets, but `processTransmit()` in `netif.cpp` reads `sTunFd` into a `1280`-byte buffer (`kMaxIp6Size = OPENTHREAD_CONFIG_IP6_MAX_DATAGRAM_LENGTH`, which includes the 4-byte BSD tunnel header!). Packets with `aLength > OT_IP6_MAX_DATAGRAM_LENGTH` (and `4 + 1280` unless `netif.cpp`'s buffer is increased by 4 bytes) will be truncated on `read(sTunFd, ...)` and dropped by `Ip6::SendRaw`. Reject packets exceeding Thread IPv6 MTU before `mDedupCache` and `mBackboneToThreadLimiter`.
2. Also reject packets with a Thread Mesh-Local source prefix (`otThreadGetMeshLocalPrefix(gInstance)`).

File: src/posix/platform/multicast_forwarding.cpp [hunk 14]
```diff
@@ -0,0 +1,321 @@
...
+        match           = slot;
+        match->mAddress = aGroup;
+        match->mInUse   = true;
+        match->mBucket.Init(mConfig.mPerGroupRatePerSecond, mConfig.mPerGroupBurst, aNow);
+    }
+
+    match->mLastUse = aNow;
+
+    // The group first: a group over its own limit must not drain the budget shared with the others.
+    VerifyOrExit(match->mBucket.Take(aNow));
```
Comment (jwhui) [inline src/posix/platform/multicast_forwarding.cpp]: If `match->mBucket.Take(aNow)` succeeds and `mTotal.Take(aNow)` fails, the per-group bucket loses a token for a dropped packet. Check/take without deducting from `mBucket` when `mTotal` is exhausted.

File: src/posix/platform/bpf_tap.hpp (hunk 1 above)
Comment (clara-mostly-human) [reply src/posix/platform/bpf_tap.hpp]: Done: `bpf_tap.hpp` and `bpf_tap.cpp` are guarded with the same condition as `multicast_routing_bpf.cpp`.

File: src/posix/platform/bpf_tap.cpp (hunk 2 above)
Comment (clara-mostly-human) [reply src/posix/platform/bpf_tap.cpp]: Done, both: the device is opened with `O_NONBLOCK`, `/dev/bpf` is tried first and the unit loop is the fallback.

File: src/posix/platform/bpf_tap.cpp (hunk 3 above)
Comment (clara-mostly-human) [reply src/posix/platform/bpf_tap.cpp]: You are right about FreeBSD, but on macOS the tunnel tap needs the "complete" write. In xnu's `bpfwrite()` (bsd/net/bpf.c) `bd_hdrcmplt` selects `dlil_output(ifp, 0, m, NULL, NULL, DLIL_OUTPUT_FLAGS_RAW, NULL)`, i.e. the frame goes out as written; a plain write goes through `dlil_output(ifp, PF_INET, m, NULL, SA(dst_buf), DLIL_OUTPUT_FLAGS_NONE, NULL)`, the normal output path with the interface's framer. So it is set on macOS for both link types, and on other platforms only for `DLT_EN10MB`, with that explanation in the comment. (FreeBSD's current `tunoutput()` also accepts `pseudo_AF_HDRCMPLT` and reads the family from `sa_data`, so newer FreeBSD would work either way.)

File: src/posix/platform/multicast_forwarding.hpp (hunk 4 above)
Comment (clara-mostly-human) [reply src/posix/platform/multicast_forwarding.hpp]: Done: the guard and the posix config include are gone; the module builds on Linux and is covered by a new gtest, `ot-posix-gtest` (`tests/gtest/multicast_forwarding_test.cpp`, 15 tests).

File: src/posix/platform/multicast_forwarding.cpp (hunk 5 above)
Comment (clara-mostly-human) [reply src/posix/platform/multicast_forwarding.cpp]: Done (see above).

File: src/posix/platform/multicast_forwarding.cpp (hunk 6 above)
Comment (clara-mostly-human) [reply src/posix/platform/multicast_forwarding.cpp]: Done: the refill is skipped for a rate of 0, and covered by `TokenBucket.ZeroRateNeverRefills`.

File: src/posix/platform/bpf_tap.cpp (hunk 7 above)
Comment (clara-mostly-human) [reply src/posix/platform/bpf_tap.cpp]: Done, with one correction to the suggested check: `bh_hdrlen >= sizeof(header)` does not hold on macOS. `sizeof(struct bpf_hdr)` is 20 here (`timeval32` plus 2 bytes of tail padding), while the header the kernel writes is padded so that the *frame* behind it is word-aligned: 18 bytes on Ethernet (14 + 18 = 32), 20 on a tunnel (4 + 20 = 24). With `>= sizeof(header)` every Ethernet record was rejected. The check is against the 18 bytes the fields occupy (`offsetof(bh_hdrlen) + sizeof(u_short)`); `bh_caplen == bh_datalen` and `<= UINT16_MAX` are checked before the cast. The buffer walk is now `BpfTap::ParseBuffer()`, tested with 18- and 20-byte headers, a zero header length, a caplen beyond the buffer and a truncated capture (`ot-posix-bpf-gtest`, built where BPF exists).

File: src/posix/platform/multicast_forwarding.cpp (hunk 8 above)
Comment (clara-mostly-human) [reply src/posix/platform/multicast_forwarding.cpp:56]: Done: `Check()` outputs the packet length (header plus payload length) and everything after it, the fingerprint, the frame and the write, uses that; loopback sources are rejected. The acceptance run injects frames padded past the packet and sees them arrive on the mesh trimmed, once each.

File: src/posix/platform/multicast_routing_bpf.cpp (hunk 9 above)
Comment (clara-mostly-human) [reply src/posix/platform/multicast_routing_bpf.cpp]: Done: one `Add()`/`Remove()` in `multicast_routing.cpp`, with the MFC calls under `#ifdef __linux__`.

File: src/posix/platform/multicast_routing_bpf.cpp (hunk 10 above)
Comment (clara-mostly-human) [reply src/posix/platform/multicast_routing_bpf.cpp]: Done: the report timeout is only set when the counters changed.

File: src/posix/platform/multicast_routing_bpf.cpp (hunk 11 above)
Comment (clara-mostly-human) [reply src/posix/platform/multicast_routing_bpf.cpp]: Done: only `OT_ERROR_FAILED` tears the taps down; `OT_ERROR_PARSE` drops the buffer with a log line and no `strerror()`. A parse error on one tap no longer skips the read of the other.

File: src/posix/platform/multicast_routing_bpf.cpp (hunk 12 above)
Comment (clara-mostly-human) [reply src/posix/platform/multicast_routing_bpf.cpp]: Done, both directions: `Check()` takes the mesh-local prefix and rejects sources in it. The length checks in `Check()` and the MTU check run before the duplicate cache and the rate limiter, so a rejected packet costs nothing.

File: src/posix/platform/multicast_routing_bpf.cpp (hunk 13 above)
Comment (clara-mostly-human) [reply src/posix/platform/multicast_routing_bpf.cpp]: Done: packets longer than `OPENTHREAD_CONFIG_IP6_MAX_DATAGRAM_LENGTH` are rejected before the caches (the acceptance run injects 1300-byte groupcasts and sees them counted as rejected, with a normal packet after them delivered). And thank you for the pointer at `processTransmit()`: its buffer is `kMaxIp6Size + 4` on the BSD platforms now, as `processReceive()`'s already was; that is in this PR since the forwarder is what made it visible, but I can split it out if you prefer.

File: src/posix/platform/multicast_forwarding.cpp (hunk 14 above)
Comment (clara-mostly-human) [reply src/posix/platform/multicast_forwarding.cpp]: Done: both buckets are checked (`HasToken()`) before either is charged; covered by `RateLimiter.GroupIsNotChargedWhenTheSharedBudgetIsExhausted`.

File: tests/gtest/bpf_tap_test.cpp [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+               uint16_t              aHeaderLength,
+               uint32_t              aCapLen,
+               uint32_t              aDataLen,
+               uint8_t               aFill)
+{
+    size_t start = aBuffer.size();
+    size_t total = BPF_WORDALIGN(aHeaderLength + aCapLen);
+
+    aBuffer.resize(start + total, 0);
+    memcpy(&aBuffer[start + 8], &aCapLen, sizeof(aCapLen));
```
Comment (jwhui) [inline tests/gtest/bpf_tap_test.cpp:63]: `AddRecord` (and the tests below) hardcodes offsets `8`, `12`, `16` and header length `18`, which assumes `bh_tstamp` is 8 bytes (`struct timeval32` on macOS). On 64-bit FreeBSD/NetBSD/OpenBSD, `bh_tstamp` is `struct timeval` (16 bytes), so `offsetof(struct bpf_hdr, bh_hdrlen)` is `24` and `kMinHeaderLength` in `BpfTap::ParseBuffer` is `26`—causing these tests to fail with `OT_ERROR_PARSE` on 64-bit BSDs.

Use `offsetof(struct bpf_hdr, ...)` and `kMinHeaderLength` so the test matches `BpfTap::ParseBuffer` across both macOS and BSD:

```cpp
constexpr uint16_t kMinHeaderLength =
     static_cast<uint16_t>(offsetof(struct bpf_hdr, bh_hdrlen) + sizeof(u_short));
```
and in `AddRecord`:
```cpp
    memcpy(&aBuffer[start + offsetof(struct bpf_hdr, bh_caplen)], &aCapLen, sizeof(aCapLen));
    memcpy(&aBuffer[start + offsetof(struct bpf_hdr, bh_datalen)], &aDataLen, sizeof(aDataLen));
    memcpy(&aBuffer[start + offsetof(struct bpf_hdr, bh_hdrlen)], &aHeaderLength, sizeof(aHeaderLength));
```
(and replace `18`, `20`, and `buffer[16..17]` in lines 76–125 with `kMinHeaderLength`, `kMinHeaderLength + 2`, and `offsetof(struct bpf_hdr, bh_hdrlen)`).

File: tests/gtest/CMakeLists.txt [hunk 16]
```diff
@@ -67,6 +67,47 @@ target_link_libraries(ot-ftd-gtest
...
+target_link_libraries(ot-posix-gtest
+    ot-fake-ftd
+    GTest::gtest
+    GTest::gtest_main
+)
+
+gtest_discover_tests(ot-posix-gtest)
+
+# The BPF tap exists only where BPF does, in a Backbone Router build: the parser test is built there.
+if(NOT CMAKE_SYSTEM_NAME STREQUAL "Linux" AND OT_BACKBONE_ROUTER)
```
Comment (jwhui) [inline tests/gtest/CMakeLists.txt:90]: `OT_BACKBONE_ROUTER` is not enabled in default `simulation` preset or `unit-tests` CI GTest builds, so `ot-posix-bpf-gtest` is currently skipped unless `-DOT_BACKBONE_ROUTER=ON` is explicitly passed.

Consider removing `AND OT_BACKBONE_ROUTER` and defining the guards directly on `ot-posix-bpf-gtest` so it always runs on non-Linux GTest builds:

```cmake
if(NOT CMAKE_SYSTEM_NAME STREQUAL "Linux")
    add_executable(ot-posix-bpf-gtest
        bpf_tap_test.cpp
        ${openthread_SOURCE_DIR}/src/posix/platform/bpf_tap.cpp
    )

    target_compile_definitions(ot-posix-bpf-gtest PRIVATE
        OPENTHREAD_CONFIG_BACKBONE_ROUTER_ENABLE=1
        OPENTHREAD_CONFIG_BACKBONE_ROUTER_MULTICAST_ROUTING_ENABLE=1
        OPENTHREAD_POSIX_CONFIG_BACKBONE_ROUTER_MULTICAST_ROUTING_ENABLE=1
    )
```

File: src/posix/platform/bpf_tap.cpp [hunk 17]
```diff
@@ -0,0 +1,245 @@
...
+{
+    // The kernel pads the record header so that the frame behind it is word-aligned: `bh_hdrlen` is the fields
+    // (18 bytes) plus that padding, and can be smaller than `sizeof(struct bpf_hdr)`, which has its own tail padding
+    // (macOS: 18 on Ethernet, 20 on a tunnel).
+    constexpr size_t kMinHeaderLength = offsetof(struct bpf_hdr, bh_hdrlen) + sizeof(u_short);
+
+    otError error  = OT_ERROR_NONE;
+    size_t  offset = 0;
+
+    while (offset + kMinHeaderLength <= aLength)
```
Comment (jwhui) [inline src/posix/platform/bpf_tap.cpp:206]: Since `offset += BPF_WORDALIGN(...)` on line 224 can advance `offset` up to `BPF_ALIGNMENT - 1` bytes past `aLength`, writing the loop condition with subtraction avoids any potential `offset + kMinHeaderLength` unsigned wrap-around on 32-bit targets:

```suggestion
    while (offset <= aLength && aLength - offset >= kMinHeaderLength)
```

File: src/posix/platform/multicast_routing_bpf.cpp [hunk 18]
```diff
@@ -0,0 +1,463 @@
...
+        mState         = kStateEnabling;
+        mNextRetryTime = otPlatTimeGet() + static_cast<uint64_t>(mRetryIntervalMs) * OT_US_PER_MS;
+        ExitNow();
+    }
+    else if (error == OT_ERROR_PARSE)
+    {
+        LogWarn("Dropped a malformed packet tap buffer");
+    }
+
+    if (otPlatTimeGet() >= mNextReportTime)
```
Comment (jwhui) [inline src/posix/platform/multicast_routing_bpf.cpp:203]: When the router has been idle for `>= kReportIntervalSec` (`mCountersChanged == false`), `mNextReportTime` is in the past. As soon as a single packet is processed in `mThreadTap.Read()` / `mBackboneTap.Read()`, `mCountersChanged` becomes `true` and this check immediately logs `ReportCounters()` for that 1 packet instead of aggregating over a `kReportIntervalSec` window (and if `Process()` runs while `mCountersChanged == false`, `ReportCounters()` still advances `mNextReportTime`).

Consider starting `mNextReportTime = otPlatTimeGet() + kReportIntervalSec * OT_US_PER_S` when `mCountersChanged` transitions from `false` to `true`, and checking `if (mCountersChanged && otPlatTimeGet() >= mNextReportTime)` here.
